### ISSUE TITLE
User survey

### ISSUE THREAD

**Issue by hickford** [hickford/git-credential-oauth#29]
Hi users. A quick survey:

1. How did you discover git-credential-oauth?
2. How did you install git-credential-oauth?
3. Any ideas for improvements?

**Comment by kra-mo** [comment]
1. `dnf search search git-credential`
2. dnf
3. I may be stupid, but I have no idea how I'm supposed to use the `get` command, some more documentation on that would be appreciated

**Comment by hickford** [comment]
@kra-mo The `get` command is called by Git. The only command you need to run is `git credential-oauth configure`.

I'll try to make this clearer.

**Comment by edavidaja** [comment]
1. I saw it go by in a list of scoop or brew updates
2. scoop / brew / into a container image
3. I will be recommending this project for use in server-based development environments once the work on headless mode is complete.

Thank you for your work on this!

**Comment by hickford** [comment]
@gzzchh  Thanks.

>    close browser tab when authorize success 

I'd like that, but AFAIK it's not technically possible because "Scripts may not close windows that were not opened by script" https://developer.mozilla.org/en-US/docs/Web/API/Window/close

**Comment by rbelem** [comment]
1. On [askubuntu.com](https://askubuntu.com/a/1469736)
2. [devbox](https://www.jetpack.io/devbox/) global add git-credential-oauth
3. On-premise bitbucket and gitlab

**Comment by Christoph-Raab** [comment]
1. Recommendation by colleague
2. Downloaded from github (not packed in ubuntu 22.04)
3. Only use it for on-prem GitLab
5 New ssh solution was horrible implemented and too much effort to setup
6 Include in ubuntu 22.04 package manager, maybe autorefresh tokens in the background, +1 to autoclose browser window

**Comment by hickford** [comment]
@Christoph-Raab Thanks for your response.

> autorefresh tokens in the background

Are you having to reauthenticate regularly in browser? What's your helper setup `git config --get-all credential.helper`? If you are using cache, try setting a longer timeout such as `cache --timeout 72000` (20 hours).

**Comment by Christoph-Raab** [comment]
I regular have to retry on a push or fetch

> remote: HTTP Basic: Access denied. The provided password or token is incorrect or your account has 2FA enabled and you must use a personal access token instead of a password. See https://<gitlab-host>/help/topics/git/troubleshooting_git#error-on-git-fetch-http-basic-access-denied

`git config --get-all credential.helper` doesn't return anything, but my git config looks like this

```
 [credential "<gitlab-host>"]
         oauthClientId = <some id>
         oauthClientSecret = <some secret>
         oauthScopes = read_repository write_repository
         oauthAuthURL = /oauth/authorize
         oauthTokenURL = /oauth/token
         helper = cache --timeout 36000
         helper = oauth
```

**Comment by hickford** [comment]
> I regularly have to retry on a push or fetch

Sounds like Git is using expired OAuth tokens. @Christoph-Raab Which version of Git are you using (`git --version`)? credential-cache has support for password_expiry_utc attribute from 2.40 (https://github.com/hickford/git-credential-oauth/issues/20) and oauth_refresh_token attribute from 2.41.

**Comment by jupblb** [comment]
1. [Home Manager](https://github.com/nix-community/home-manager) news
2. [Home Manager](https://github.com/nix-community/home-manager)
3. No
4. `cache --timeout 36000`
5. No longer need to deploy my private ssh key
6. All good, this is a great tool!

**Comment by Svenlaa** [comment]
1. [Stackoverflow answer](https://stackoverflow.com/a/76345423)
2. sudo dnf install git-credential-oauth
3. Yes GitLab, but not this OS/Machine
4. ```cache --timeout 7200 \n oauth```
5. I wanted to clone my private repo from github on linux, and this got my creds setup
6. A survey where all answers are public maybe isn't the best idea

**Comment by abh** [comment]
1. Looking for how to get `go get ...` to authenticate to a private https-only git repository (where I don't think I actually needed this; disabling GOSUMDB for my domain was the solution).
2. Homebrew
3. Yes, Gitea (and gitosis and GitHub enterprise and ...)
4. ```% git config --get-all credential.helper
osxkeychain
cache --timeout 10400
oauth``` plus
```
[credential "https://gitea...."]
        helper = oauth
	oauthClientId = a386d92d-2983-....
	oauthScopes = read_repository
	oauthAuthURL = /login/oauth/authorize
	oauthTokenURL = /login/oauth/access_token
```
5. Less long lived magic tokens.
6. I'll make a separate issue

**Comment by voltagex** [comment]
>How did you discover git-credential-oauth?

Searched, because it's 2024 and Git on Linux still doesn't have password storage by default

>How did you install git-credential-oauth?

deb package, then uninstalled as it pulled in most of an XServer, then realised it wouldn't work for me as it's a remote box and I'm not SSH forwarding

>Do you authenticate to any Git hosts outside GitHub?

Yes

>What's your credential.helper configuration? Run git config --get-all credential.helper to check.
Nothing, on this box, apparently.

"manager" on Windows.


>What problem did git-credential-Oauth solve for you?

None yet.

>Any ideas for improvements?

I don't know - without CLI-only OAuth working this won't work for me

**Comment by hickford** [comment]
@voltagex 

> deb package pulled in most of an XServer

Debian package [git-credential-oauth](https://packages.debian.org/stable/git-credential-oauth) has no dependencies. It recommends package xdg-open, but you can avoid installing it with `apt install --no-install-recommends git-credential-oauth`.

> it wouldn't work for me as it's a remote box

Have you tried `helper = oauth -device`? This works on any system including browserless systems. https://github.com/hickford/git-credential-oauth/blob/main/README.md#browserless-systems

**Comment by voltagex** [comment]
Thanks for your help @hickford

I am able to use it on my box that has a UI and a web browser, but `-device` with Forgejo leads to `endpoint missing DeviceAuthURL` which I believe needs some changes from Forgejo itself.

**Comment by j-lakeman** [comment]
>     1. How did you discover git-credential-oauth?
`apt search git credential` / Stack Overflow
>     2. How did you install git-credential-oauth?
`apt` 
>     3. Do you authenticate to any Git hosts outside GitHub?

- Gitea
- Forgejo
- Codeberg
- GitLab

>     6. Any ideas for improvements?

pls support more “indie” platforms like the ones above

**Comment by midsorbet** [comment]
> Have you tried `helper = oauth -device` (v0.11.0 or later)? This works on any system including browserless systems. https://github.com/hickford/git-credential-oauth/blob/main/README.md#browserless-systems
GitHub kept giving me a link that would redirect back to `127.0.0.1:37653` after logging in from a browser. I had to setup a SSH tunnel to set it up on my hetzner box (`ssh -i -L 127.0.0.1:37653:127.0.0.1:37653  hetzner`). It did work as advertised when I was setting it up on my local machine.

System details
OS: nixOS 24.05
Version: 0.11.3

Git config
```
[credential]
        helper = "/nix/store/xl2hmm62sbjfm574car3a8ac45y83qik-git-credential-oauth-0.11.3/bin/git-credential-oauth"
        helper = "cache --timeout 120"
        helper = "oauth -device"
```

**Comment by mass8326** [comment]
1. Discovered git-credential-oath when Googling for a seamless way of authenticating like on Windows
2. Previously installed using the PPA, but now I install using Go
3. No hosts used outside GitHub
4. See below for my credential.helper configuration
5. git-credential-oauth prevents the annoyance of having to use SSH or PATs
6. I've opened a separate issue for a potential improvement
 
My credential.helper on bare metal Linux:
```
/usr/share/doc/git/contrib/credential/libsecret/git-credential-libsecret
oauth
```
My credential.helper on WSL:
```
cache --timeout 21600
oauth
```

**Comment by JDLH** [comment]
> 1. How did you discover git-credential-oauth?

I was working on a private GitHub repository where I needed my local computer's bare git to authenticate to GitHub using OAuth. (The project did not want to issue personal access tokens or use ssh.) [GitHub's git cli instructions](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git#git-credential-manager) recommended using Git Credential Manager, installing via Homebrew. I use MacPorts and not Homebrew, so I searched MacPorts for a git credential helper supporting oauth. I found this.

> 2. How did you install git-credential-oauth?

Using MacPorts and its [**git-credential-oauth** port](https://ports.macports.org/port/git-credential-oauth/).

> 3. Do you authenticate to any Git hosts outside GitHub?

No. And not GitHub yet! I am having difficulty.

> 4. What's your credential.helper configuration? Run git config --get-all credential.helper to check.

```
% git config --get-all credential.helper                   
osxkeychain
oauth -verbose
```

> 5. What problem did git-credential-Oauth solve for you?

I hope that it will let me work with a private repo on GitHub. 

> 6. Any ideas for improvements?

Better diagnostics for my failure case.

**Comment by mariuszste** [comment]
1. stack overflow
2. AUR
3. yes, forgejo
4. 
```
cache --timeout 21600
oauth
```
5. not being able to use SSH, no passwords on disk, "modern"
6. Document what's supported/tested. I was shocked when it worked first try. A support matrix telling you what works out of the box, what requires additional configuration, what doesn't and why, etc.

**Comment by Lockszmith-GH** [comment]
> * How did you discover git-credential-oauth?

[forgego's documentation](https://forgejo.org/docs/latest/user/oauth2-provider/#git-authentication).

> * How did you install git-credential-oauth?

I install it using chezmoi's externals support - it downloads it directly from github/release:

<details> <summary>The <b>git-credentials-oauth</b> section of my <code>.chezmoiexternal.yaml</code> file: - click to expand</summary>

```yaml
# https://www.chezmoi.io/reference/special-files-and-directories/chezmoiexternal-format/
...
{{ $defaultRefresh := "168h" -}}
{{- $externals := dict -}}
...
{{- with $repo := "hickford/git-credential-oauth" -}}
{{-   with $version := trimPrefix "v" (gitHubLatestRelease $repo).GetTagName -}}
{{-     $_ := set $externals ".local/bin/git-credential-oauth" (dict
          "type" "archive-file"
          "url" (printf "https://github.com/%s/releases/download/v%s/git-credential-oauth_%s_linux_amd64.tar.gz" $repo $version $version)
          "path" "git-credential-oauth"
          "executable" true
          "refreshPeriod" $defaultRefresh
) -}}
{{-   end -}}
{{- end -}}
...
{{- $externals | toYaml }}
...
```
</details>

> * Do you authenticate to any Git hosts outside GitHub?

My own Forgejo instance

> * What's your credential.helper configuration? Run `git config --get-all credential.helper` to check.

```ini
...
[credential]
  helper = cache --timeout 21600
  helper = oauth
[credential "https://<my forgejo fqdn>"]
  oauthClientId = a4792ccc-144e-407e-86c9-5e7d8d9c3269
  oauthAuthURL = /login/oauth/authorize
  oauthTokenURL = /login/oauth/access_token
...
```

> * What problem did git-credential-Oauth solve for you?

As designed, no need for token generation, using web browser to authenticate (I copy-paste the URL into the browser on my Windows machine, I then `curl` the resulting URL from within the Linux machine I'm running the git cli on a parallel shell - I use `zellij` to easily run in parallel)

> * Any ideas for improvements?

Allow to specify a different hostname in the `redirect_uri`, `127.0.0.1` is great when I'm working locally, but if I'm on the LAN connected to another machine, I have to copy-paste the redirect URL from the response in the browser and run curl locally to approve the handshake, especially when there is no `-device` support, like the case with Forgejo at the moment.

**Comment by hinricht** [comment]
How did you discover git-credential-oauth?

Probably through https://docs.gitlab.com/user/profile/account/two_factor_authentication/#git-credential-oauth

How did you install git-credential-oauth?

Through the AUR package: `pamac install git-credential-oauth`

Do you authenticate to any Git hosts outside GitHub?

Yes 

What's your credential.helper configuration? Run git config --get-all credential.helper to check.

```sh
$ git config --get-all credential.helper
cache --timeout 604800
oauth 
```

What problem did git-credential-Oauth solve for you?

* Don't need to rotate any PATs anymore

Any ideas for improvements?

* Right now, I need to authorize **each repository** once until the cache timeout is reached. While this is ok in my normal daily routine, it doesn't scale when you want to use tools like gitlabber which recurively clone all repos of an instance you have access to. Is there any chance to allow all repos for such an operation ?

**Comment by hinricht** [comment]
When comparing to [git-credential-manager (GCM)](https://github.com/git-ecosystem/git-credential-manager), I found out that this behaviour is because `git-credential-oauth` stores **one secret per repo** in the git credential store, whereas GCM only stores **one secret per git instance** (i.e. one secret for `gitlab.com`.
Is this the right observation ? If so it would be nice to document this difference in the README, because it doesn't scale well with many repositories.

**Comment by Lockszmith-GH** [comment]
> * How did you discover git-credential-oauth?

Searching for a lightweight alternative to GCM, as I wanted tha functionality on a machine I couldn't install .net on.
I believe a post on SO led me here.

> * How did you install git-credential-oauth?

I grab the binary and configure .gitconfig.
This is automated via chezmoi.

> * Do you authenticate to any Git hosts outside GitHub?
Yes, I use it with GitLab at work, and my personal Forgejo instance at home.

> * What's your credential.helper configuration? Run `git config --get-all credential.helper` to check.

```
cache --timeout 21600
oauth
```
  
> * What problem did git-credential-Oauth solve for you?

No more typing user-name/passwords, using long passwords makes sense, I now rely on my vaultwarden/bitwarden for keeping my secrets in the browser, and everything is much more secure with TOTP.
 
> * Any ideas for improvements?

1. Enable (per host) an "auto retry once" when git auth fails, so re-invokation is more automatic.
2. A copy-paste workflow when device isn't available would make OAuth on remote ssh sessions flows easier.

   * SSH into a remote machine
   * When it's time to git push, prompt comes up with URL
   * Copy URL into my local browser, and complete the login workflow
   * Copy the response URL from the browser page
   * Paste into terminal <--- I currently open a new shell and run curl with the URL

**Comment by lovelymono** [comment]
> 1. How did you discover git-credential-oauth?

<https://git-scm.com/docs/gitcredentials#_oauth>

> 2. How did you install git-credential-oauth?

NixOS through home-manager.

> 3. Do you authenticate to any Git hosts outside GitHub?

GitLab.

> 4. What's your credential.helper configuration? Run `git config --get-all credential.helper` to check.

```
cache --timeout 21600
/nix/store/7mc144lbia5wkfp0qs91bld61xkj4l1c-git-credential-oauth-0.17.2/bin/git-credential-oauth
```

> 5. What problem did git-credential-Oauth solve for you?

Easier alternative to using Personal Access Tokens.

> 6. Any ideas for improvements?

Seconding Devin, closing the tab automaticaly after a timeout would be nice.
